Improvements

reginleif888 opened the pull request.

reginleif888 committed:
depositFromAllowance() improvements

reginleif888 committed:
Update example according to icrc-84 changes

reginleif888 committed:
Remove debug comments from tests

reginleif888 committed:
Fix LogEvent type

reginleif888 committed:
Add test for depositFromAllowance() when caller principal != account owner

reginleif888 committed:
Add notify_tag_commit_hook.yml

reginleif888 committed:
Fix tests

timohanke commented on src/AccountManager.mo:
```diff
@@ -360,40 +365,41 @@ module {
       let transferResult = await* processAllowance(p, account, amount);
 
-      let originalCredit : Nat = amount - fee(#deposit);
+      let originalCredit : Nat = amount - ledgerFee_;
```
According to the spec here https://github.com/dfinity/ICRC-1/tree/main/standards/ICRC-2#icrc2_transfer_from  `transfer_from` adds the ledger fee to the `amount` value so that the `to` account gets exactly `amount`. Therefore here `originalCredit = amount`. No subtraction of `ledgerFee_`.

timohanke commented on src/AccountManager.mo:
```diff
@@ -360,40 +365,41 @@ module {
         case (#Err(#BadFee { expected_fee })) {
           updateFee(expected_fee);
-          let originalCredit_2 : Nat = Int.abs(Int.min(originalCredit, amount - fee(#deposit)));
+          let originalCredit_2 : Nat = Int.abs(Int.min(originalCredit, amount - ledgerFee_));
```
Same as above. `originalCredit_2` is probably not needed. We just retry everything exactly as before after calling `updateFee`.

timohanke commented on src/AccountManager.mo:
```diff
@@ -360,40 +365,41 @@ module {
             };
             case (#Err err) {
-              log(p, #consolidationError(err));
+              log(p, #depositViaAllowanceError(err));
```
Maybe `#allowanceError` is enough? just shorter

timohanke reviewed (commented):
Left some comments

reginleif888 committed:
Fix deposit via allowance

timohanke reviewed (commented):
LGTM.

Wondering if we should keep some of the deleted tests? There were tests for when the ledger fee changes while deposit from allowance is underway. Problems with the credit calculation are no longer possible so that would indeed justify removing the tests. But we are still retrying in the BadFee branch. So the retry itself would still get covered by those tests. Shall we therefore keep them or some of them?

timohanke commented on src/AccountManager.mo:
```diff
@@ -360,40 +365,38 @@ module {
+      if (amount <= ledgerFee_) {
+        log(p, #allowanceError(#TooLowQuantity));
+        return #err(#TooLowQuantity);
+      };
```
We should remove this check and just test if amount is > 0.
The reason is that sender pays the fee. 

We also have another check inside `processAllowance` so this one should be superfluous.

timohanke replied:
Resolved in subsequent refactoring.

timohanke committed:
remove TooLowQuantity check in depositFromAllowance

timohanke committed:
Refactor BadFee retries, error types

For withdrawals and deposit by allowances we have immediate retries in case of
BadFee. We now do these retries by recursion. This removes nested switch
statements and some code repetition.

Result variants are switched to the Motoko native lower case #ok and #err. This
allows us to use the standard Result package more frequently.

We also introduce more type definitions for internal use to shorten some code
lines.

timohanke committed:
Wrap a class around ledger api and abstract away the async calls

We define three calls for consolidation, drawing from allowance,
and sending out from main account.
